### DISCUSSION TITLE
How to run doctr on M1 Mac

### DISCUSSION THREAD

**Discussion by adesgautam** [mindee/doctr#749] [Q&A] [answered]
Hi,

I am currently trying to run the "crnn_vgg16_bn" on M1 Mac. 
But doctr seems to be incompatible with the tensorflow-macos library.

Can you please tell how to use doctr on M1 Mac ?

**Comment by adesgautam** [comment, marked as answer]
I figured out to run it somehow, try the following:

* Install Anaconda x86_64 from [here](https://repo.anaconda.com/archive/Anaconda3-2021.11-MacOSX-x86_64.pkg) 
```
Anaconda3-2021.11-MacOSX-x86_64.pkg
```
It will ask you to install Rosetta2, just allow it.

* Open terminal & create an env using the same python version: 
```
conda create --name tfmac python=3.8.5
conda activate tfmac
```

* Install custom tensorflow whl file (whl file link to download: https://drive.google.com/drive/folders/1oSipZLnoeQB0Awz8U68KYeCPsULy_dQ7):
```
pip install tensorflow-2.4.1-py3-none-any.whl
```

* Install doctr & dependencies:
```
pip install python-doctr==0.4.1
pip install tensorflow_addons==0.15.0
```

* Install Cairo & Pango (only for doctr):
```
conda install -c anaconda cairo
conda install -c anaconda pango
```

The whl file described above is from [here](https://github.com/tensorflow/tensorflow/issues/46044#issuecomment-828347288)
All of this will run libs for x86_64 & not native arm64 on M1, therefore, it will probably be slow.

Hope this work for you  and please do tell if it works

**Reply by fg-mindee** [reply]
Hi @adesgautam :wave: 

Thanks a lot! As you mentioned, the M1 chip issues are upstream of docTR from its own dependencies, so we have limited options to help users with that. Some frameworks have publicly said they have started on working on M1 chip support, so hopefully that will require fewer tweaks in the future!

Much appreciated that you investigated on your own and managed to find a solution. By any chance, are you able to run the sample analysis script? (`python scripts/analyze.py /path/to/your/image`)

**Reply by adesgautam** [reply]
Hi @fg-mindee 

```
python scripts/analyze.py test.png
```

I ran the script on this [image](https://templates.invoicehome.com/invoice-template-us-classic-white-750px.png) and the output was perfect.

Below is the output for the script you mentioned:

```
/Users/adeshgautam/opt/anaconda3/envs/tfmac/lib/python3.8/site-packages/tensorflow_addons/utils/ensure_tf_install.py:53: UserWarning: Tensorflow Addons supports using Python ops for all Tensorflow versions above or equal to 2.5.0 and strictly below 2.8.0 (nightly versions are not supported). 
 The versions of TensorFlow you are currently using is 2.4.1 and is not supported. 
Some things might work, some things might not.
If you were to encounter a bug, do not file an issue.
If you want to make sure you're using a tested and supported configuration, either change the TensorFlow version or the TensorFlow Addons's version. 
You can find the compatibility matrix in TensorFlow Addon's readme:
https://github.com/tensorflow/addons
  warnings.warn(
WARNING:tensorflow:Unresolved object in checkpoint: (root).fpn.layer_blocks.0.layer_with_weights-0._inbound_nodes
WARNING:tensorflow:Unresolved object in checkpoint: (root).fpn.layer_blocks.0.layer_with_weights-1._inbound_nodes
WARNING:tensorflow:Unresolved object in checkpoint: (root).fpn.layer_blocks.0.layer-2._inbound_nodes
WARNING:tensorflow:Unresolved object in checkpoint: (root).fpn.layer_blocks.1.layer_with_weights-0._inbound_nodes
WARNING:tensorflow:Unresolved object in checkpoint: (root).fpn.layer_blocks.1.layer_with_weights-1._inbound_nodes
WARNING:tensorflow:Unresolved object in checkpoint: (root).fpn.layer_blocks.1.layer-2._inbound_nodes
WARNING:tensorflow:Unresolved object in checkpoint: (root).fpn.layer_blocks.1.layer-3._inbound_nodes
WARNING:tensorflow:Unresolved object in checkpoint: (root).fpn.layer_blocks.2.layer_with_weights-0._inbound_nodes
WARNING:tensorflow:Unresolved object in checkpoint: (root).fpn.layer_blocks.2.layer_with_weights-1._inbound_nodes
WARNING:tensorflow:Unresolved object in checkpoint: (root).fpn.layer_blocks.2.layer-2._inbound_nodes
WARNING:tensorflow:Unresolved object in checkpoint: (root).fpn.layer_blocks.2.layer-3._inbound_nodes
WARNING:tensorflow:Unresolved object in checkpoint: (root).fpn.layer_blocks.3.layer_with_weights-0._inbound_nodes
WARNING:tensorflow:Unresolved object in checkpoint: (root).fpn.layer_blocks.3.layer_with_weights-1._inbound_nodes
WARNING:tensorflow:Unresolved object in checkpoint: (root).fpn.layer_blocks.3.layer-2._inbound_nodes
WARNING:tensorflow:Unresolved object in checkpoint: (root).fpn.layer_blocks.3.layer-3._inbound_nodes
WARNING:tensorflow:A checkpoint was restored (e.g. tf.train.Checkpoint.restore or tf.keras.Model.load_weights) but not all checkpointed values were used. See above for specific issues. Use expect_partial() on the load status object, e.g. tf.train.Checkpoint.restore(...).expect_partial(), to silence these warnings, or use assert_consumed() to make the check explicit. See https://www.tensorflow.org/guide/checkpoint#loading_mechanics for details.
WARNING:tensorflow:Unresolved object in checkpoint: (root).decoder.layer_with_weights-0._inbound_nodes
WARNING:tensorflow:Unresolved object in checkpoint: (root).decoder.layer_with_weights-1._inbound_nodes
WARNING:tensorflow:Unresolved object in checkpoint: (root).decoder.layer_with_weights-2._inbound_nodes
WARNING:tensorflow:A checkpoint was restored (e.g. tf.train.Checkpoint.restore or tf.keras.Model.load_weights) but not all checkpointed values were used. See above for specific issues. Use expect_partial() on the load status object, e.g. tf.train.Checkpoint.restore(...).expect_partial(), to silence these warnings, or use assert_consumed() to make the check explicit. See https://www.tensorflow.org/guide/checkpoint#loading_mechanics for details.
```

**Reply by ghost** [reply]
Hello there 👋🏼 ,

I have just installed `python-doctr` in a 🍎 M1 Macbook 🍎 :

```bash
$ python collect_env.py
Collecting environment information...

DocTR version: 0.5.1
TensorFlow version: 2.10.0
PyTorch version: 1.13.0.dev20220927 (torchvision 0.14.0.dev20220927)
OpenCV version: 4.6.0
OS: Mac OSX 12.6
Python version: 3.10.6
Is CUDA available (TensorFlow): Yes
Is CUDA available (PyTorch): No
CUDA runtime version: Could not collect
GPU models and configuration: /physical_device:GPU:0
Nvidia driver version: Could not collect
cuDNN version: Could not collect
```

First, I followed the [Apple instructions to install `tensorflow-metal`](https://developer.apple.com/metal/tensorflow-plugin/). Then, as @adesgautam [mentioned](https://github.com/mindee/doctr/discussions/749#discussioncomment-1867776), I installed `python-doctr` and `tensorflow-addons` following these steps:

1. `python -m pip install python-doctr`
2. `python -m pip install tensorflow-addons`

(I didn't need to specify any version, but others might have to)

4. and also had to install `cairo` and `pango`:

> Install Cairo & Pango (only for doctr):
> ```
> conda install -c anaconda cairo
> conda install -c anaconda pango
>```

I finished all these steps by installing PyTorch:

1. `conda install pytorch torchvision torchaudio -c pytorch`
2. `python -m pip install "python-doctr[torch]"`  

and I could finally [get my pretrained model](https://github.com/mindee/doctr#getting-your-pretrained-model) 😝 

Hope it helps 🙏🏼 

> **Note**: To avoid `fontconfig` warning, just run `conda install fontconfig` in your environment.

---

> By any chance, are you able to run the sample analysis script? (`python scripts/analyze.py /path/to/your/image`)

I ran the `analyse.py` script successfully 😄🎉  

```
$ python scripts/analyze.py ../data/example.png 
Metal device set to: Apple M1 Pro

systemMemory: 32.00 GB
maxCacheSize: 10.67 GB

DEBUG:tensorflow:Layer lstm will use cuDNN kernels when running on GPU.
DEBUG:tensorflow:Layer lstm will use cuDNN kernels when running on GPU.
DEBUG:tensorflow:Layer lstm will use cuDNN kernels when running on GPU.
DEBUG:tensorflow:Layer lstm_1 will use cuDNN kernels when running on GPU.
DEBUG:tensorflow:Layer lstm_1 will use cuDNN kernels when running on GPU.
DEBUG:tensorflow:Layer lstm_1 will use cuDNN kernels when running on GPU.
```

**Reply by felixdittrich92** [reply]
Hi @DRTorresRuiz 👋 , really nice 👍 
Thanks a lot for sharing your success of setting up doctr on M1 :)
